usr/src/uts/common/io/usb/clients/ccid/ccid.c

```diff
@@ -0,0 +1,4426 @@
+	if (cmp == NULL)
+		return;
+	cv_signal(&cmp->cm_excl_cv);
+}
+
+static void
+ccid_slot_excl_rele(ccid_slot_t *slot)
+{
+	ccid_minor_t *cmp;
+	ccid_t *ccid = slot->cs_ccid;
+
+	VERIFY(MUTEX_HELD(&ccid->ccid_mutex));
+	VERIFY3P(slot->cs_excl_minor, !=, NULL);
+
+	cmp = slot->cs_excl_minor;
+
+	/*
+	 * If we have an outstanding command left by the user when they've
+	 * closed the slot, we need to clean up this command. We need to call
+	 * the protocol specific handler here to determine what to do. If the
+	 * command has completed, but the user has never called read, then it
+	 * will simply clean it up. Otherwise it will indicate that there is
+	 * some amount of external state still ongoing to take care of and clean
+	 * up later.
+	 */
+	if (slot->cs_icc.icc_teardown != NULL) {
+		slot->cs_icc.icc_teardown(ccid, slot, ECANCELED);
+	}
+
+	/*
+	 * There may either be a thread blocked in read or in the process of
+	 * preparing a write. In either case, we need to make sure that they're
+	 * woken up or finish, before we finish tear down.
+	 */
+	while ((cmp->cm_flags & CCID_MINOR_F_READ_WAITING) != 0 ||
+	    (slot->cs_io.ci_flags & CCID_IO_F_PREPARING) != 0) {
+		cv_wait(&cmp->cm_iowait_cv, &ccid->ccid_mutex);
+	}
+
+	/*
+	 * At this point, we hold the lock and there should be no other threads
+	 * that are past the basic sanity checks. So at this point, note that
+	 * this minor no longer has exclusive access (causing other read/write
+	 * calls to fail) and start the process of cleaning up the outstanding
+	 * I/O on the slot. It is OK that at this point the thread may try to
+	 * obtain exclusive access again. It will end up blocking on everything
+	 * else.
+	 */
+	cmp->cm_flags &= ~CCID_MINOR_F_HAS_EXCL;
+	slot->cs_excl_minor = NULL;
+
+	/*
+	 * If at this point, we have an I/O that's noted as being done, but no
+	 * one blocked in read, then we need to clean that up. The ICC teardown
+	 * function is only designed to take care of in-flight I/Os.
+	 */
+	if ((slot->cs_io.ci_flags & CCID_IO_F_DONE) != 0)
+		ccid_clear_io(&slot->cs_io);
+
+	/*
+	 * Regardless of when we're polling, we need to go through and error
+	 * out.
+	 */
+	pollwakeup(&cmp->cm_pollhead, POLLERR);
+
+	/*
+	 * If we've been asked to reset the card before handing it off, schedule
+	 * that. Otherwise, allow the next entry in the queue to get woken up
+	 * and given access to the card.
+	 */
+	if ((cmp->cm_flags & CCID_MINOR_F_TXN_RESET) != 0) {
+		slot->cs_flags |= CCID_SLOT_F_NEED_TXN_RESET;
+		ccid_worker_request(ccid);
+		cmp->cm_flags &= ~CCID_MINOR_F_TXN_RESET;
+	} else {
+		ccid_slot_excl_maybe_signal(slot);
+	}
+}
+
+static int
+ccid_slot_excl_req(ccid_slot_t *slot, ccid_minor_t *cmp, boolean_t nosleep)
+{
+	VERIFY(MUTEX_HELD(&slot->cs_ccid->ccid_mutex));
+
+	if (slot->cs_excl_minor == cmp) {
+		VERIFY((cmp->cm_flags & CCID_MINOR_F_HAS_EXCL) != 0);
+		return (EEXIST);
+	}
+
+	if ((cmp->cm_flags & CCID_MINOR_F_WAITING) != 0) {
```

Comment:
This can happen if an open minor is shared by threads or processes, the driver will still only have one ccid_minor_t for that and doesn't know or care about the actual threads or processes. If both threads owning a single ccid_minor_t attempt to get exclusive access at the same time, one of them will get EINPROGRESS.

I think it would be preferable if we didn't have to worry about this at all and just disallow such use cases. This would be a nice case for O_CLOFORK, too.

Reply:
Couldn't we just get rid of the flag and allow threads to queue up on the CV? Losers might have to re-add 'cmp' to the list again but that seems OK? Is this for when we're trying to tear a slot down maybe?

Alternatively, if we really want to make it impossible for threads to wait on a txn like this, we should at least clearly comment this here - and I can't remember what the manpage says but it should do so there if not.

What do you think?

Reply:
We already handle multiple threads waiting for exclusive access if they all have their own cloned minor. I'm not sure it's justified to complicate this driver more by allowing multiple threads to wait on each other on the same cloned minor. It just seems as a rather obscure use case, and I fail to see what benefit it would give to an application.

I'll check the manpage to see if there's a need to explain this behavior better.